py4j-python/src/py4j/java_gateway.py

```diff
@@ -678,6 +760,39 @@ def __getattr__(self, name):
         # The name is a method
         return self._methods[name]
 
+    def __dir__(self):
+        self._populate_fields()
+        return list(set(self._methods.keys()) | self._field_names)
+
+    def _populate_fields(self):
+        if not self._fully_populated:
```

Comment:
I would add a note similar to the one in __getattr__ stating that this is not thread safe but the worst case scenario is that this method is called twice (at the same time) and the _methods dict is filled twice with the same values. We may want to make everything thread-safe (or thread-efficient :-) ) later and this comment will be very helpful.